snuba_sdk/expressions.py

```diff
@@ -130,44 +130,163 @@ class Function(Expression):
+        if func.alias is not None:
+            if not isinstance(func.alias, str) or func.alias == "":
                 raise InvalidExpression(
-                    f"alias '{self.alias}' of function {self.function} must be None or a non-empty string"
+                    f"alias '{func.alias}' of function {func.function} must be None or a non-empty string"
                 )
-            if not column_name_re.match(self.alias):
+            if not self.column_name_re.match(func.alias):
                 raise InvalidExpression(
-                    f"alias '{self.alias}' of function {self.function} contains invalid characters"
+                    f"alias '{func.alias}' of function {func.function} contains invalid characters"
                 )
 
-        for param in self.parameters:
+        for param in func.parameters:
             if isinstance(param, (Column, Function, *Scalar)):
                 continue
             else:
                 assert not isinstance(param, bytes)  # mypy
                 raise InvalidExpression(
-                    f"parameter '{param}' of function {self.function} is an invalid type"
+                    f"parameter '{param}' of function {func.function} is an invalid type"
                 )
 
-    def translate(self) -> str:
-        alias = "" if self.alias is None else f" AS {self.alias}"
+        return True
+
+    def visit_int_literal(
+        self, literal: int, minn: Optional[int], maxn: Optional[int], name: str
+    ) -> bool:
+        if not isinstance(literal, int):
+            raise InvalidExpression(f"{name} '{literal}' must be an integer")
+        if minn is not None and literal < minn:
+            raise InvalidExpression(f"{name} '{literal}' must be at least {minn:,}")
+        elif maxn is not None and literal > maxn:
+            raise InvalidExpression(f"{name} '{literal}' is capped at {maxn:,}")
+
+        return True
+
+    def visit_entity(self, entity: Entity) -> bool:
+        # TODO: There should be a whitelist of entity names at some point
+        if not self.entity_name_re.match(entity.name):
+            raise InvalidEntity(f"{entity.name} is not a valid entity name")
+
+        return True
+
+    def visit_condition(self, cond: Condition) -> bool:
+        if not isinstance(cond.lhs, (Column, Function)):
+            raise InvalidExpression(
+                f"invalid condition: LHS of a condition must be a Column or Function, not {type(cond.lhs)}"
+            )
+        if not isinstance(cond.rhs, (Column, Function, *Scalar)):
+            raise InvalidExpression(
+                f"invalid condition: RHS of a condition must be a Column, Function or Scalar not {type(cond.rhs)}"
+            )
+        if not isinstance(cond.op, Op):
+            raise InvalidExpression(
+                "invalid condition: operator of a condition must be an Op"
+            )
+
+        return True
+
+
+class Translation(Visitor[str]):
```

Comment:
Did you keep the visitor in the same module for a specific reason? While the Visitor abstract class needs to stay in the same module as the visited class with this implementation, the implementations of the visitor do not have the same constraint and can be anywhere.

Reply:
Not quite, because the base Expression class uses both the Validation and Translation visitors to automatically validate/translate. I could probably rewrite this to move the translation one out of this file, but the validation one would need to stay.

Reply:
Right... meh.
Let's take a step back. The reason we started talking about visitors (or pattern matching if you prefer that alternative) is to decouple a feature that needs to traverse the tree from the core node classes themselves. This allows you not to squeeze all those responsibilities into the core classes themselves, which would soon explode in complexity.
Now it seems this works well for translators, as you can have multiple translation mechanisms (think about pretty print vs simple compact format vs formatting for tracing that removes literals). And the fact you can move that class out of this module seems a good sign.

If validation is inherently coupled with the core classes because you call validation in post_init, decoupling it is not that relevant. The question is another one though: how much logic do you expect to add in the post_init validation ? So far it seems simple enough that extracting from the core classes or not is not a major change. But I remember we talked about contextual validation. Would that still be tightly coupled with the core classes or would it be separate ?

Reply:
We can keep contextual validation at the Query level for now, or maybe as separate visitors. So I think the validation on the individual expressions should remain fairly static. I'll remove the Validation visitor, that should clean this up quite a bit.

Reply:
Make sense to me, though I think we may need to decide sooner than later what kind of contextual validation we want to implement and whether we allow the client to plug their own validation logic (in could be interesting to let the client define dataset specific validation, but I did not put a lot of thinking in it).